### PR TITLE
feat(tooltip): Added key value tooltip feature

### PR DESCRIPTION
**Pull request by jbrundage**
Also added a use-case example to Graph.les_miz
Signed-off-by: Jaman Brundage <jbrundage372@gmail.com>

### CODE REVIEW

**Comment by jbrundage** [comment]
image

File: packages/api/src/ITooltip.ts [hunk 1]
```diff
@@ -123,6 +123,20 @@ export abstract class ITooltip extends Widget {
         }
     }
 
+    tooltipKeyValueFormat(obj) {
+        const _key_color = this.tooltipLabelColor() ? this.tooltipLabelColor() : this.tooltipSeriesColor();
```
Comment (GordonSmith) [inline packages/api/src/ITooltip.ts]: Should use `tooltipLabelColor_exists` or it will fail on color "0"

File: packages/graph/src/Graph.ts [hunk 2]
```diff
@@ -448,10 +448,42 @@ export class Graph extends SVGZoomWidget {
...
+                return "";
+            });
+        }
+    }
+
     update(domNode, element) {
         super.update(domNode, element);
         this.tooltip.hide();
 
+        this.setTooltipType();
```
Comment (GordonSmith) [inline packages/graph/src/Graph.ts]: Can't do this here as the user may have already provided their own tooltip function.

File: packages/graph/src/Graph.ts [hunk 3]
```diff
@@ -448,10 +448,42 @@ export class Graph extends SVGZoomWidget {
...
+        } else {
+            this.tooltipHTML(function (d: any) {
+                if (d instanceof Subgraph) {
+                    return `<p style="text-align:center">${d.title().replace("\n", "<br>")}</p>`;
+                } else if (d instanceof Vertex) {
+                    return `<p style="text-align:center">${d.text().replace("\n", "<br>")}</p>`;
+                } else if (d instanceof Edge) {
+                    return `<p style="text-align:center">${d.text().replace("\n", "<br>")}</p>`;
+                }
+                return "";
```
Comment (GordonSmith) [inline packages/graph/src/Graph.ts]: This is duplicate code...

File: packages/graph/src/Graph.ts [hunk 4]
```diff
@@ -448,10 +448,42 @@ export class Graph extends SVGZoomWidget {
...
+                    return `<p style="text-align:center">${d.title().replace("\n", "<br>")}</p>`;
+                } else if (d instanceof Vertex) {
+                    return context.tooltipKeyValueFormat({
+                        label: d.text().replace("\n", "<br>"),
+                        connections: context.data().edges.filter(e => e.targetVertex().id() === d.id() || e.sourceVertex().id() === d.id()).length
+                    });
+                } else if (d instanceof Edge) {
+                    return `<p style="text-align:center">${d.text().replace("\n", "<br>")}</p>`;
+                }
+                return "";
```
Comment (GordonSmith) [inline packages/graph/src/Graph.ts]: You could make this the default option in the constructor?

File: packages/graph/src/Graph.ts [hunk 5]
```diff
@@ -944,6 +977,11 @@ export class Graph extends SVGZoomWidget {
 Graph.prototype._class += " graph_Graph";
 Graph.prototype.implements(IGraph.prototype);
 Graph.prototype.implements(ITooltip.prototype);
+export interface Graph {
+    useDetailedTooltip(): boolean;
+    useDetailedTooltip(_: boolean): this;
+}
+Graph.prototype.publish("useDetailedTooltip", false, "boolean");
```
Comment (GordonSmith) [inline packages/graph/src/Graph.ts]: This paradigm doesn't work - as the whole purpose of the tooltipHtml function is to allow the user to provide their own.

File: packages/graph/src/Graph.ts [hunk 6]
```diff
@@ -448,9 +448,29 @@ export class Graph extends SVGZoomWidget {
...
+            return "";
+        });
+    }
+
     update(domNode, element) {
         super.update(domNode, element);
         this.tooltip.hide();
+        if (this.useDetailedTooltip()) {
+            this.overrideTooltipHTML();
+        }
```
Comment (jbrundage) [inline packages/graph/src/Graph.ts]: @GordonSmith I've updated it to be optional

File: packages/graph/src/Graph.ts (hunk 5 above)
Comment (GordonSmith) [reply packages/graph/src/Graph.ts]: Can you add a description and move it so its below "snapToGrid".

Also will you rebase onto latest candidate-2.x.x as I think it won't merge correctly as it stands (I am bit surprised that GitHub isn't warning about it...

**Comment by jbrundage** [comment]
@GordonSmith updated with a sample
image

**Comment by jbrundage** [comment]
BTW this sample exposed some odd BubbleXY behavior. I'll add an issue to track it.

File: packages/api/src/ITooltip.ts [hunk 7]
```diff
@@ -123,10 +123,25 @@ export abstract class ITooltip extends Widget {
...
+        switch (this.tooltipStyle()) {
+            case "none":
+                break;
+            default:
+                return Object.keys(obj).map(key => _keyValueSpans(key, obj[key])).join("<br/>");
+        }
+        function _keyValueSpans(_key, _val) {
+            return `<span style='color:${_key_color}'>${_key}</span>:  <span style='color:${_value_color}'>${_val}</span>`;
+        }
+    }
```
Comment (GordonSmith) [inline packages/api/src/ITooltip.ts]: There is a similar implementation to this here:  https://github.com/hpcc-systems/Visualization/blob/master/demos/imdb/src/IMDBVertex.ts#L70-L86 
With better alignment - I am not sure you can support the "title" feature in the general case though?

Comment (jbrundage) [reply packages/api/src/ITooltip.ts]: @GordonSmith I've updated the PR with your alternative function. It does look nicer.

**Pull request merged**